sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c

```diff
@@ -486,191 +536,147 @@ static void _send_multicast(gnrc_netif_t *netif, gnrc_pktsnip_t *pkt,
+        if ((pkt = _create_netif_hdr(NULL, 0, pkt, netif_hdr_flags)) == NULL) {
             return;
         }
     }
+    if (_safe_fill_ipv6_hdr(netif, pkt, prep_hdr)) {
+        _send_multicast_over_iface(pkt, netif, netif_hdr_flags);
+    }
+#endif  /* GNRC_NETIF_NUMOF */
+}
 
-    if (prep_hdr) {
-        if (_fill_ipv6_hdr(netif, ipv6, payload) < 0) {
-            /* error on filling up header */
-            gnrc_pktbuf_release(pkt);
-            return;
-        }
+static void _send_to_self(gnrc_pktsnip_t *pkt, bool prep_hdr,
+                          gnrc_netif_t *netif)
+{
+    uint8_t *rcv_data;
+    gnrc_pktsnip_t *ptr = pkt, *rcv_pkt;
+
+    if (!_safe_fill_ipv6_hdr(netif, pkt, prep_hdr)) {
+        return;
     }
+    rcv_pkt = gnrc_pktbuf_add(NULL, NULL, gnrc_pkt_len(pkt), GNRC_NETTYPE_IPV6);
 
-    _send_multicast_over_iface(netif, pkt);
-#endif  /* GNRC_NETIF_NUMOF */
+    if (rcv_pkt == NULL) {
+        DEBUG("ipv6: error on generating loopback packet\n");
+        gnrc_pktbuf_release(pkt);
+        return;
+    }
+
+    rcv_data = rcv_pkt->data;
+
+    /* "reverse" packet (by making it one snip as if received from NIC) */
+    while (ptr != NULL) {
+        memcpy(rcv_data, ptr->data, ptr->size);
+        rcv_data += ptr->size;
+        ptr = ptr->next;
+    }
+
+    gnrc_pktbuf_release(pkt);
+
+    DEBUG("ipv6: packet is addressed to myself => loopback\n");
+
+    if (gnrc_netapi_receive(gnrc_ipv6_pid, rcv_pkt) < 1) {
+        DEBUG("ipv6: unable to deliver packet\n");
+        gnrc_pktbuf_release(rcv_pkt);
+    }
 }
 
 static void _send(gnrc_pktsnip_t *pkt, bool prep_hdr)
 {
     gnrc_netif_t *netif = NULL;
-    gnrc_pktsnip_t *ipv6, *payload;
-    ipv6_hdr_t *hdr;
+    gnrc_pktsnip_t *tmp_pkt;
+    ipv6_hdr_t *ipv6_hdr;
+    uint8_t netif_hdr_flags = 0U;
+
     /* get IPv6 snip and (if present) generic interface header */
     if (pkt->type == GNRC_NETTYPE_NETIF) {
         /* If there is already a netif header (routing protocols and
-         * neighbor discovery might add them to preset sending interface) */
+         * neighbor discovery might add them to preset sending interface or
+         * higher layers wants to provide flags to the interface ) */
+        const gnrc_netif_hdr_t *netif_hdr = pkt->data;
+
         netif = gnrc_netif_get_by_pid(((gnrc_netif_hdr_t *)pkt->data)->if_pid);
-        /* seize payload as temporary variable */
-        ipv6 = gnrc_pktbuf_start_write(pkt); /* write protect for later removal
-                                              * in _send_unicast() */
-        if (ipv6 == NULL) {
+        /* discard broadcast and multicast flags because those could be
+         * potentially wrong (dst is later checked to assure that multicast is
+         * set if dst is a multicast address) */
+        netif_hdr_flags = netif_hdr->flags &
+                          ~(GNRC_NETIF_HDR_FLAGS_BROADCAST |
+                            GNRC_NETIF_HDR_FLAGS_MULTICAST);
```

Comment:
Is there a place where these are set, and if so, why if they are not used anyway. Or is this just a safe guard against them being set?

Reply:
They are set in [`_create_netif_hdr()`](https://github.com/RIOT-OS/RIOT/pull/8215/files#diff-74976d45a42eaa063e3561b5fd36f615R358) again. Or do you mean if they may be set somewhere else? I'm not sure this is the case now, but there was the plan to use them for markers, if the checksum for UDP header compression can be elided, e.g.